### PR TITLE
Return bad request on too many surrogate ids

### PR DESCRIPTION
**Pull request by SergeyGaluzo**
When >80K resources with the same last updated are ingested, code fails with
Current - 500
After fix - 400 with good message

fixes https://microsofthealth.visualstudio.com/Health/_workitems/edit/120027/

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 1]
```diff
@@ -444,10 +444,10 @@ async Task<(List<ImportResource> Loaded, List<ImportResource> Conflicts)> Import
                 }
                 else if (importMode == ImportMode.IncrementalLoad)
                 {
-                    // dedup by last updated - keep all versions when version and lastUpdated exist on record.
+                    // dedup by last updated - take first version for single last updated, prefer large version.
                     var inputsDedupped = resources
-                        .GroupBy(_ => _.ResourceWrapper.ToResourceDateKey(_model.GetResourceTypeId, ignoreVersion: !_.KeepVersion || !_.KeepLastUpdated))
-                        .Select(_ => _.First())
+                        .GroupBy(_ => _.ResourceWrapper.ToResourceDateKey(_model.GetResourceTypeId, ignoreVersion: true))
+                        .Select(_ => _.OrderByDescending(_ => _.ResourceWrapper.Version).First())
```
Comment (mikaelweave) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:450]: Is this a change of behavior? Will this no longer allow customers to have multiple versions with the same last updated?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/ImportProcessingJob.cs [hunk 2]
```diff
@@ -31,6 +33,7 @@ public class ImportProcessingJob : IJob
     {
         private const string CancelledErrorMessage = "Import processing job is canceled.";
         internal const string DefaultCallerAgent = "Microsoft.Health.Fhir.Server";
+        internal const string SurrogateIdsErrorMessage = "Unable to generate internal IDs. If the lastUpdated meta element is provided as input, reduce the number of resources with the same up to millisecond lastUpdated below 10,000.";
```
Comment (mikaelweave) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/ImportProcessingJob.cs:36]: NIT - consider resource file if appropriate. Ignore if not.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 3]
```diff
@@ -459,8 +459,8 @@ async Task<(List<ImportResource> Loaded, List<ImportResource> Conflicts)> Import
 
                 async Task HandleIncrementalVersionedImport(List<ImportResource> inputs, bool useReplicasForReads)
                 {
-                    // Dedup by version via ToResourceKey - only keep first occurance of a version in this batch.
-                    var inputsWithVersion = inputs.Where(_ => _.KeepVersion).GroupBy(_ => _.ResourceWrapper.ToResourceKey()).Select(_ => _.First()).ToList();
+                    // Dedup by version via ToResourceKey - prefer latest dates.
+                    var inputsWithVersion = inputs.Where(_ => _.KeepVersion).GroupBy(_ => _.ResourceWrapper.ToResourceKey()).Select(_ => _.OrderByDescending(_ => _.ResourceWrapper.LastModified.DateTime).First()).ToList();
```
Comment (mikaelweave) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:463]: Nice - I like this change. Makes more sense than "first in file". 🦾

**Review by mikaelweave — COMMENTED** [review]
Some questions but in general PR looks good.

Will approve after hearing back about change near:
`// dedup by last updated - take first version for single last updated, prefer large version.`

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 1 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:450]: Yes. It is a fix for a bug. This only affects cases with incorrect input data.

**Review by mikaelweave — APPROVED** [review]
🦾

**Pull request merged**